### DISCUSSION TITLE
Generating private key (as string) from generated Uint8Array

### DISCUSSION THREAD

**Discussion by HopiumArtist** [paulmillr/noble-hashes#28] [Q&A] [answered]
Hi,

I am using your libraries as follows to generate the Uint8Array:

```js
const mn = bip39.generateMnemonic(wordlist);
const seed = await bip39.mnemonicToSeed(mn);
```

I am curious how I get proceed further to also generate the private key as a string.

I tried this trying to imitate the README https://github.com/paulmillr/noble-hashes#usage

```js
const privateKey = bytesToHex(sha256(seed))
```

Any guidance appreciated!

Thanks for your awesome work.

**Comment by paulmillr** [comment, marked as answer]
I assume you're talking about secp256k1. secp256k1 can receive 32 bytes as private key. It can directly work with Uint8Arrays. `mnemonicToSeed` produces 64 bytes of output. Which means you can strip its 32 bytes:

```ts
const mn = bip39.generateMnemonic(wordlist);
const seed = await bip39.mnemonicToSeed(mn);
const privateKey = seed.slice(0, 32);
const publicKey = secp256k1.getPublicKey(privateKey);
```

There is a caveat: not all 32-byte values are not valid private keys. E.g. `"ff".repeat(32)` is not a valid key becasuse it's bigger than `CURVE.n`. You can modulo divide it to get a valid key in these rare circumstances:

```ts
const mn = bip39.generateMnemonic(wordlist);
const seed = await bip39.mnemonicToSeed(mn);
const privateKeyN = secp256k1.utils.mod(BigInt('0x' + bytesToHex(seed)))); // number
const privateKey = privateKeyN.toString(16).padStart(64, '0'); // hex string
const publicKey = secp256k1.getPublicKey(privateKey);
```

If you're talking about ed25519, any 32-bit string/uint8array is fine - no need to do utils.mod.

**Comment by HopiumArtist** [comment]
thank you!
